Sources/SwiftBSON/BSONDocument.swift

```diff
@@ -517,3 +517,42 @@ extension BSONDocument: BSONValue {
 extension BSONDocument: CustomStringConvertible {
     public var description: String { self.toExtendedJSONString() }
 }
+
+extension BSONDocument {
+    /// Helper function to check for BSON document equality regardless of the order of key-value pairs
+    internal func equalsIgnoreOrderHelper(a: BSONDocument, b: BSONDocument) -> Bool {
+        let keysA = a.keys.sorted()
+        let keysB = b.keys.sorted()
+
+        guard keysA == keysB else {
+            return false
+        }
+
+        for k in keysA {
+            if case let (.document(aDoc), .document(bDoc)?) = (a[k], b[k]) {
+                guard self.equalsIgnoreOrderHelper(a: aDoc, b: bDoc) else {
+                    return false
+                }
+                continue
+            }
+            guard a[k] == b[k] else {
+                return false
+            }
+        }
+
+        return true
+    }
+
+    /**
+     * Compares this `BSONDocument` with another provided one and returns true if they have the same key/value pairs
```

Comment:
"Compares this `BSONDocument` with another" and "returns true if they have" are a little redundant; I think we can just drop them without losing any clarity: "Returns whether this `BSONDocument` contains exactly the same key/value pairs as the provided `BSONDocument`, regardless of the order of the keys". 

I think we should also include a warning that this method is much less efficient than checking for regular equality since the document is internally ordered.